### DISCUSSION TITLE
perfetto UI can see streaming data directly from linux target?

### DISCUSSION THREAD

**Discussion by deco-dev** [google/perfetto#946] [Q&A] [answered]
Hello.

Environment
Host: Windows amd64(Chrome)
Target: Linux arm64(Ubuntu)
Interface: USB host - device or USB RNDIS.

I want to see profiling data to realtime on perfetto UI. But perfetto UI only supports ADB device.
But my target board is operating Linux arm64(Ununtu). So I can not use adbd.
I tried to this.
In host PC I runs trace_processor_shell -D. and enter perfetto UI.
In target board I runs traced, traced_probes and perfetto -txt -c config.json -o -.
But the trace data just is printed in STDOUT.

Please help me.

**Comment by LalitMaganti** [comment, marked as answer]
There is no "easy & out of the box" support for streaming data over the network/USB from non-Android devices with Perfetto.  If you are willing to get your hands dirty, you might want to checkout traced_relay [1] which is designed for forwarding tracing data across VMs but hypothetically should also be able to forward tracing data over the network across machines. However, note that this is completely untested and we **cannot** provide support for this. You are on your own to figure out this stuff if you go this route.

Also just want to clarify that when collecting traces on Android, the data on the UI is not displayed "relatime" like you seem to suggest. it's just streamed to the host where it's stored in a file/in memory and then when tracing is completed the file is parsed as a whole and then opened.

> In host PC I runs trace_processor_shell -D. and enter perfetto UI.

No idea why running trace_processor_shell here is mentioned, it's not relevant at all really.

> But the trace data just is printed in STDOUT.

Which makes sense because you have specified `-o -` which means "write the trace to the command line".

[1] https://github.com/google/perfetto/blob/2cb2b19bcf55f1923ad78a64b692d397a322c7ae/src/traced_relay/relay_service_main.cc#L31

**Comment by deco-dev** [comment]
Then is it difficult to see profiling in realtime of Android devices using perfetto UI in host PC?

**Reply by LalitMaganti** [reply]
There's no real time profiling visualization available on android.

**Reply by LalitMaganti** [reply]
It's one of those features which sounds very cool but in practice it's not a huge deal. It's also immensely complex to actually implement because trace information is written out of order and is heterogeneous. Making all the visualization work correctly is so hard that it's basically not worth it.

**Reply by deco-dev** [reply]
OK. I understood.

**Reply by deco-dev** [reply]
Other profiling Tools (like ARM Development Sudio or Qualcomm Profiler) support live profiling(realtime).
So I tried to it using perfetto ui.

**Reply by LalitMaganti** [reply]
Sure but Perfetto supports many more event types and heterogeneous data which those tools do not bother dealing with. If those tools work better for you, then you should feel free to use them :)

**Comment by deco-dev** [comment]
I have one more question.
perfetto supports hw_perf_event like cache-misses, cpu-cycles? I tested it, it did not work well.

My goal is hw perf event(pmu) and trace data(like ftrace, proc/, sys stat) show visualization at same time on realtime.

**Reply by LalitMaganti** [reply]
Yes it does but it's quite difficult to configure correctly. You need to make sure you are using the MONOTONIC clock as the default boottime clock does not work.

Quoting from an internal config which is trying to collect cache refill events:
```
data_sources: {
  config {
    name: "linux.perf"
    perf_event_config {
      timebase {
        period: 100
        raw_event {
          type: 8 # cat /sys/bus/event_source/devices/armv8_pmuv3/type ==> 8
          config: 23 # cat /sys/bus/event_source/devices/armv8_pmuv3/events/l2d_cache_refill =>> event=0x0017
        }
        timestamp_clock: PERF_CLOCK_MONOTONIC
        name: "l2d_cache_refill"
      }
    }
  }
}
```

**Reply by deco-dev** [reply]
Oh.. So diffcult. I see the reference of perfetto, https://perfetto.dev/docs/reference/trace-config-proto
It is difference. Why does it set raw_event? In the reference, counter uses enum values.
https://perfetto.dev/docs/reference/trace-config-proto#PerfEvents.Counter

**Reply by LalitMaganti** [reply]
You can use the enum values if they work for you: cache refills are very hardware specific and do not have a generic interface at the perf level: raw events are an escape hatch for such events.

If the counter enums work for you, you should use them. The important part is the presence of `timestamp_clock: PERF_CLOCK_MONOTONIC`

**Reply by deco-dev** [reply]
OK. I will try it!

**Reply by deco-dev** [reply]
I used data sources like this. But no trace data.
```
buffers: {
  size_kb: 300000
  fill_policy: DISCARD
}

data_sources: {
  config {
    name: "linux.perf"
    perf_event_config {
      timebase {
        period: 100
        counter: HW_CACHE_MISSES
        timestamp_clock: PERF_CLOCK_MONOTONIC
      }
    }
  }
}
duration_ms: 10000
```

**Reply by LalitMaganti** [reply]
How are you collecting the trace? I guess you are on Linux: are you using tracebox or are you running daemons manually?

**Reply by deco-dev** [reply]
Right. I am on linux. I used traced and traced_probes to background mode and I used perfetto.

**Reply by LalitMaganti** [reply]
You also require traced_perf to be running for perf sampling. It's a separate Daemon if you are building from sources or you can just get it all in one if you run tracebox.

**Reply by deco-dev** [reply]
ok got it. If I get trace data to use tracebox, then I can see the data on perfetto UI?

**Reply by LalitMaganti** [reply]
It should work from everything I know :)

**Reply by deco-dev** [reply]
It worked well! Thank you!
